torch/csrc/jit/script/compiler.cpp

```diff
@@ -36,115 +36,153 @@ using TypeTable = std::unordered_map<std::string, TypePtr>;
+    // if the result is false, both a and b had to be false,
+    // so we take their union.
+    return RefinementSet(
+        intersectList(true_refinements_, rhs.true_refinements_),
+        unionList(false_refinements_, rhs.false_refinements_));
   }
 
-  c10::optional<TypeAndRange> getRefinement(const std::string& name) const {
-    const auto& maybe_mapping = mappings_.find(name);
-    if (maybe_mapping == mappings_.end()) {
-      return c10::nullopt;
-    }
-    return maybe_mapping->second;
-  }
-
-  // return the intersection of the values to type mappings between this
-  // types can be unified
-  void intersectRefinements(const Refinements& other) {
-    Refinements ret;
-    for (const auto& name_mapping : mappings_) {
-      const auto& name = name_mapping.first;
-      const auto& mapping = name_mapping.second;
-      if (auto other_mapping = other.getRefinement(name_mapping.first)) {
-        const auto maybe_unified_type =
-            unifyTypes(mapping.first, other_mapping->first);
-        if (maybe_unified_type) {
-          ret.setRefinement(
-              name, TypeAndRange(*maybe_unified_type, mapping.second));
-        }
-      }
-    }
-    mappings_ = std::move(ret.mappings_);
-  }
-
-  // return the union of the values to type mappings in a and b whose
-  // types can be unified
-  void unionRefinements(const Refinements& other) {
-    Refinements ret;
-    for (const auto& name_mapping : mappings_) {
-      const auto& name = name_mapping.first;
-      const auto& mapping = name_mapping.second;
-      TypePtr t_1 = mapping.first;
-      if (auto other_mapping = other.getRefinement(name_mapping.first)) {
-        TypePtr t_2 = other_mapping->first;
-        c10::optional<TypePtr> maybe_unified_type = c10::nullopt;
-        if (t_1->isSubtypeOf(t_2)) {
-          maybe_unified_type = t_1;
-        } else if (t_2->isSubtypeOf(t_1)) {
-          maybe_unified_type = t_2;
-        }
-        if (maybe_unified_type) {
-          ret.setRefinement(
-              name, TypeAndRange(*maybe_unified_type, mapping.second));
-        }
-      } else {
-        ret.setRefinement(name, mapping);
+  RefinementSet Not() const {
+    return RefinementSet(false_refinements_, true_refinements_);
+  }
+  const std::vector<Refinement> activeRefinements() const {
+    return true_refinements_;
+  }
+
+ private:
+  static Refinements unionList(const Refinements& a, const Refinements& b) {
+    Refinements result = a;
+    for (const Refinement& r : b) {
+      auto it = std::find(a.begin(), a.end(), r);
+      if (it == a.end()) {
+        result.push_back(r);
       }
     }
+    return result;
+  }
 
-    for (auto& name_mapping : other.mappings_) {
-      if (!getRefinement(name_mapping.first)) {
-        ret.setRefinement(name_mapping.first, name_mapping.second);
+  static Refinements intersectList(const Refinements& a, const Refinements& b) {
+    Refinements result;
+    for (const Refinement& r : a) {
```

Comment:
What is the motivation is for removing the existing `intersection ` and `union` implementations? The new implementation probably suffices for just optional refinement, but with the possibility of more complicated refinement using `isinstance` and an `Any` type i think we should use the more correct existing implementation. 

Something like `isinstance(x, float) or isinstance(x, int)` would just accept the second type instead of unifying them.  

Note: the existing `intersection` and `union` produces the correct refinements, just the `insertRefinements` function incorrectly inserts a refinement when the refinement is not None.

Reply:
discussed in person: there may be unexpected consequences of unifyTypes refinement. For instance if we derive a type Scalar, it is another instance of injecting a Scalar in addition to the item call. This has led to surprising bugs in the past, so we should have a good example case before we make more liberal use of refinement. The behavior currently implemented is a subset of the valid refinements that we would do in the future, so it is safe to expand.